contracts/src/v0.8/vrf/VRFCoordinatorV2Plus.sol

```diff
@@ -0,0 +1,611 @@
+    // during the consumers callback code via reentrancyLock.
+    // Note that callWithExactGas will revert if we do not have sufficient gas
+    // to give the callee their requested amount.
+    bool success = callWithExactGas(rc.callbackGasLimit, rc.sender, resp);
+
+    // stack too deep error
+    {
+      // We want to charge users exactly for how much gas they use in their callback.
+      // The gasAfterPaymentCalculation is meant to cover these additional operations where we
+      // decrement the subscription balance and increment the oracles withdrawable balance.
+      uint96 payment = calculatePaymentAmount(
+        startGas,
+        s_config.gasAfterPaymentCalculation,
+        tx.gasprice,
+        rc.nativePayment
+      );
+      if (rc.nativePayment) {
+        if (s_subscriptions[rc.subId].ethBalance < payment) {
+          revert InsufficientBalance();
+        }
+        s_subscriptions[rc.subId].ethBalance -= payment;
+        s_withdrawableEth[s_provingKeys[output.keyHash]] += payment;
+      } else {
+        if (s_subscriptions[rc.subId].balance < payment) {
+          revert InsufficientBalance();
+        }
+        s_subscriptions[rc.subId].balance -= payment;
+        s_withdrawableTokens[s_provingKeys[output.keyHash]] += payment;
+      }
+
+      // Include payment in the event for tracking costs.
+      // event RandomWordsFulfilled(uint256 indexed requestId, uint256 outputSeed, uint96 payment, bool nativePayment, bool success);
+      emit RandomWordsFulfilled(output.requestId, output.randomness, payment, rc.nativePayment, success);
+
+      // Update the payment field in the request payments mapping so that
+      // consumers know how much they've paid and can charge end-users accordingly.
+      s_requestPayments[output.requestId] = payment;
+
+      return payment;
+    }
+  }
+
+  function calculatePaymentAmount(
+    uint256 startGas,
+    uint256 gasAfterPaymentCalculation,
+    uint256 weiPerUnitGas,
+    bool nativePayment
+  ) internal view returns (uint96) {
+    if (nativePayment) {
+      return
+        calculatePaymentAmountEth(
+          startGas,
+          gasAfterPaymentCalculation,
+          s_feeConfig.fulfillmentFlatFeeEthPPM,
+          weiPerUnitGas
+        );
+    }
+    return
+      calculatePaymentAmountLink(
+        startGas,
+        gasAfterPaymentCalculation,
+        s_feeConfig.fulfillmentFlatFeeLinkPPM,
+        weiPerUnitGas
+      );
+  }
+
+  function calculatePaymentAmountEth(
+    uint256 startGas,
+    uint256 gasAfterPaymentCalculation,
+    uint32 fulfillmentFlatFeePPM,
+    uint256 weiPerUnitGas
+  ) internal view returns (uint96) {
+    // Will return non-zero on chains that have this enabled
+    uint256 l1CostWei = ChainSpecificUtil.getCurrentTxL1GasFees();
+    // calculate the payment without the premium
+    uint256 baseFeeWei = weiPerUnitGas * (gasAfterPaymentCalculation + startGas - gasleft());
```

Comment:
Note: this and `calculatePaymentAmountLink` overlap, but we need only worry about this if the contract size gets too big.

Reply:
Agreed, there is room for refactoring here.